### ISSUE TITLE
Don't emit UnresolvableInclude after a is_file check

### ISSUE THREAD

**Issue by koriym** [vimeo/psalm#4788]
```php
<?php
function load(string $file) : void
{
    if (is_file($file)){
    	require $file;
    }
}

```
Expected: No issue
Actual: UnresolvableInclude

Related: #908 #3502
https://psalm.dev/r/d8f0074db3

`file_exists()` cause no error.
https://psalm.dev/r/bd0cf68fd7

**Comment by weirdan** [comment]
Should be easy to add here: https://github.com/vimeo/psalm/blob/cec8d7138fb983bf4be3ee6a4f5107cb82ea9f35/src/Psalm/Internal/Analyzer/Statements/Expression/Call/NamedFunctionCallHandler.php#L118-L126

**Comment by muglug** [comment]
So that's not quite what `UnresolvableInclude` is for. Instead it’s meant to tell the user that Psalm was unable to locate the file for analysis, so it's anyone's guess what the file does.

It's _not_ meant to say "this file doesn't exist".

**Comment by weirdan** [comment]
Sure, but if `file_exists()` suppresses it, why `is_file()` shouldn't?

**Comment by muglug** [comment]
Oops. Sorry. I guess https://github.com/vimeo/psalm/issues/908 changed the meaning indeed. OK then!

**Comment by kermitmadlyn** [comment]
tested with v4.6.2:

emits UnresolvableInclude:
```
if (is_file($template)) {
  require $template;
}
```

does not emit UnresolvableInclude:
```
if (file_exists($template)) {
  require $template;
}
```

emits UnresolvableInclude:
```
if (!empty($template) && file_exists($template)) {
  require $template;
}
```

**Comment by pawel-slowik** [comment]
The original issue regarding `is_file` was fixed in https://github.com/vimeo/psalm/pull/5263 .

The other reported issue:
```php
if (!empty($template) && file_exists($template)) {
    require $template;
}
```
is still unsolved. The cause is that the property used for tracking potentially existing files (`phantom_files`) isn't always passed between analyzed contexts. If I understand the code correctly, we'd have to update / copy `phantom_files` somewhere around here: https://github.com/vimeo/psalm/blob/0bb1648f4d8ec39385c2ae69fd91856cd2b08a6c/src/Psalm/Internal/Analyzer/Statements/Block/IfConditionalAnalyzer.php#L200-L201

I started fiddling with it, but I'm not sure if this is really worth the effort. Instead, I'd recommend refactoring the problematic code into guard clauses:
```php
if (empty($file)) {
    return;
}

if (!file_exists($file)) {
    return;
}

require $file;
```

What do you think?

**Comment by orklah** [comment]
Context juggling is quite tricky, I'm often confused by the number of context there are and which one contains what.

If you can fix it, it would be great as it seems there were more than one person confused by that but it's not a big priority

**Comment by pawel-slowik** [comment]
So I've finally had the time to dig into this some more, but ultimately I failed. The problem is that the current implementation only works for simple cases and I don't feel up to rewriting it. For example, this should throw `UnresolvableInclude`, but doesn't:

```php
function requireFile(string $s) : void {
    if (is_file($s)) {
        // NOP
    }
    require_once($s);
}
```

Therefore, I can't create tricky, context dependent tests. No tests, no PR.

Since currently there are no tests at all for `UnresolvableInclude`, I've added some simple cases here: https://github.com/vimeo/psalm/pull/7021 .
